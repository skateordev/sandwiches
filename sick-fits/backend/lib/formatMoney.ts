const formatter = new Intl.NumberFormat('en-US', {
  style: 'currency',
  currency: 'EUR',
});

export default function formatMoney(subunit: number) {
  const units = subunit / 100;
  return formatter.format(units);
}
