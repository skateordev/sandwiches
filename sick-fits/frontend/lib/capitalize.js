/**
 * Takes the given string and makes the first character uppercase
 * @param string: the string to capitalize
 */

function capitalize(string) {
  return string.charAt(0).toUpperCase() + string.slice(1);
}

export default capitalize;
