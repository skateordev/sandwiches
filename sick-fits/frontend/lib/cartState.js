import PropTypes from 'prop-types';
import { createContext, useContext, useState } from 'react';

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;

function CartStateProvider({ children }) {
  // this is our own custom provider!
  // we will store data (state) and functionality (updaters) in here
  // anyone can acces it via the consumer!

  const [cartOpen, setCartOpen] = useState(false);

  function openCart() {
    setCartOpen(true);
  }

  function closeCart() {
    setCartOpen(false);
  }

  return (
    <LocalStateProvider value={{ cartOpen, openCart, closeCart }}>
      {children}
    </LocalStateProvider>
  );
}

CartStateProvider.propTypes = {
  children: PropTypes.node,
};

CartStateProvider.defaultProps = {
  children: null,
};

// make a custom hook for access the cart local state
function useCart() {
  // use a consumer to access the local state
  const all = useContext(LocalStateContext);

  return all;
}

export {
  useCart,
  CartStateProvider,
};
