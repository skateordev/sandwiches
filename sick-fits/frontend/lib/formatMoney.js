export default function formatMoney(amount = 0) {
  const options = {
    style: 'currency',
    currency: 'EUR',
    minimumFractionDigits: 2,
  };

  // check if it's a whole dollar amount
  if (amount % 100 === 0) {
    options.minimumFractionDigits = 0;
  }

  // english-Eurozone (countries that fully support the euro)
  const formatter = Intl.NumberFormat('en-EZ', options);

  return formatter.format(amount / 100);
}
