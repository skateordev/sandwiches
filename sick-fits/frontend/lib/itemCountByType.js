/**
 * Determines the number of items for the given itemType
 * Used primarily for pagination and cache
 * @param {object} data: object read from cache or returnd from paginationQuery
 * @param {string} itemType
 * @returns {number} itemType meta count property
 */
import { ITEM_TYPE_ORDERS, ITEM_TYPE_PRODUCTS } from '../components/constants';

export default function itemCountByType(data, itemType) {
  let count = 0;

  switch (itemType) {
    case ITEM_TYPE_ORDERS:
      count = data?._allOrdersMeta?.count;
      break;
    case ITEM_TYPE_PRODUCTS:
      count = data?._allProductsMeta?.count;
      break;
    default:
      break;
  }

  return count;
}
