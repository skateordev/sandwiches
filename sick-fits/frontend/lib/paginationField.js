import { ITEM_TYPE_ORDERS, ITEM_TYPE_PRODUCTS } from '../components/constants';

export default function paginationField(query, itemType) {
  return ({
    keyArgs: false, // tells apollo we will take care of everything
    read: (existing = [], { args, cache }) => { // eslint-disable-line default-param-last
      /**  first we ask `read` function for some items
        *  we can either:
        *    return items already in cache
        *    - or -
        *    return false (and make/cause a network request to refresh) */

      const { skip, first } = args;

      // read number of items on the page from cache
      const data = cache.readQuery({ query });
      let count = 0;

      // determine what we're paginating
      switch (itemType) {
        case ITEM_TYPE_ORDERS:
          count = data?._allOrdersMeta?.count;
          break;
        case ITEM_TYPE_PRODUCTS:
          count = data?._allProductsMeta?.count;
          break;
        default:
          break;
      }

      // current page number and total number of pages
      const page = skip / first + 1;
      const pages = Math.ceil(count / first);

      /* check if we have existing items and also filter out undefined
         if doing 4 items per page then on the 2nd page skip would be 4
         skip + first = 8, so items = 5, 6, 7, 8 */
      const items = existing.slice(skip, skip + first).filter((item) => item);

      if (items.length && items.length !== first && page === pages) {
        /* when there aren't enough items for a full page (partial last page)
           just gimme whatever there is */

        return items;
      }

      if (items.length !== first) {
        // there are no items! go to network and fetch!

        return false;
      }

      if (items.length) {
        /* items have been found in cache, don't need to
           go to network */
        console.log(`💥 GOT EEEM!\n🔍 Found ${existing.length} cached items!\n🪂 Sending to apollo...`);

        return items;
      }

      return false; // fallback to network request
    },
    merge: (existing, incoming, { args }) => {
      // this runs when the Apollo client comes back from the network with our product
      const { skip } = args;

      console.log(`Merging ${incoming.length} items from the network 🌐`);

      const merged = existing ? existing.slice(0) : [];

      for (let i = skip; i < skip + incoming.length; i++) {
        merged[i] = incoming[i - skip];
      }

      return merged;
    },
  });
}
