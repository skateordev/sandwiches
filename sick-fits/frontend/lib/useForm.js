import { useEffect, useState } from "react"

export default function useForm(initial = {}) {
  const [inputs, setInputs] = useState(initial);

  // fix values of `initial` being undefined
  const initialValues = Object.values(initial).join('');

  useEffect(() => {
    // initialValues will be undefined on first render
    // update inputs to the actual values once initialValues changes
    setInputs(initial);
  }, [initialValues]);

  function handleChange(evt) {
    let { name, type, files, value } = evt.target;

    if (type === 'number') {
      value = parseInt(value);
    }

    if (type === 'file') {
      // assign files to the first item (aka first file) in the [value] array
      [value] = files;
    }

    setInputs({
      ...inputs,
      [name]: value,
    })
  }

  function resetForm() {
    setInputs(initial)
  }

  function clearForm() {
    console.log('inputs', inputs);
    const blankState = Object.fromEntries(
      Object.entries(inputs).map(([key, value]) => [key, ''])
    );

    setInputs(blankState);
  }

  return {
    inputs,
    clearForm,
    resetForm,
    handleChange,
  }
};
