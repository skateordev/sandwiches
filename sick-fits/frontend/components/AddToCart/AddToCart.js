import PropTypes from 'prop-types';
import { useMutation } from '@apollo/client';
import styled from 'styled-components';
import ADD_TO_CART_MUTATION from './mutations/addToCartMutation';
import CURRENT_USER_QUERY from '../User/queries/currentUserQuery';
import { useCart } from '../../lib/cartState';

const AddToCartStyled = styled.button`
  &:hover {
    background-color: var(--red);
    color: var(--white);
  }
`;

export default function AddToCart({ id }) {
  const [addToCart, { loading: isLoading }] = useMutation(ADD_TO_CART_MUTATION, {
    variables: { productId: id },
    refetchQueries: [{ query: CURRENT_USER_QUERY }],
  });

  const { openCart } = useCart();

  const handleClick = () => {
    openCart();
    addToCart();
  };

  return (
    <AddToCartStyled
      type="button"
      onClick={handleClick}
      disabled={isLoading}
      aria-busy={isLoading}
    >
      Add{isLoading && 'ing'} to cart ðŸ›’
    </AddToCartStyled>
  );
}

AddToCart.propTypes = {
  id: PropTypes.string.isRequired,
};
