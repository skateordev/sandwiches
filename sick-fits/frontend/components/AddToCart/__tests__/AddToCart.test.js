/* eslint-disable no-undef */
import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';
import userEvent from '@testing-library/user-event';
import { fakeItem } from '../../../lib/testUtils';
import { CartStateProvider } from '../../../lib/cartState';
import ADD_TO_CART_MUTATION from '../mutations/addToCartMutation';
import AddToCart from '../AddToCart';

const { id } = fakeItem();
const mocks = [
  {
    request: {
      query: ADD_TO_CART_MUTATION,
      varibles: {
        id,
      },
    },
    result: {
      data: {
        addToCart: { id },
      },
    },
  },
];

beforeEach(() => jest.resetAllMocks());

describe('<AddToCart />', () => {
  it('matches the default snapshot', () => {
    const { container } = render(
      <CartStateProvider>
        <MockedProvider>
          <AddToCart productId={id} />
        </MockedProvider>
      </CartStateProvider>,
    );

    expect(container).toMatchSnapshot();
  });

  it('add an item to the cart when clicked', () => {
    render(
      <CartStateProvider>
        <MockedProvider mocks={mocks}>
          <AddToCart productId={id} />
        </MockedProvider>
      </CartStateProvider>,
    );

    const btn = screen.getByText('Add to cart ðŸ›’');

    userEvent.click(btn);

    const btnBusy = screen.getByText('Adding to cart ðŸ›’');

    expect(btnBusy).toBeDisabled();
    expect(btnBusy).toHaveAttribute('aria-busy');
  });
});
