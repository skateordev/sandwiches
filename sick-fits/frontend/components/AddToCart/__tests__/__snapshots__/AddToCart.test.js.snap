// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddToCart /> matches the default snapshot 1`] = `
<div>
  <button
    aria-busy="false"
    class="AddToCart__AddToCartStyled-sc-6bl65o-0 fHiDfE"
    type="button"
  >
    Add
     to cart ðŸ›’
  </button>
</div>
`;
